common/src/comments.rs

```diff
@@ -2,91 +2,296 @@ use crate::{
+        (**self).take_leading(pos)
     }
 
-    pub fn leading_comments(
-        &self,
-        pos: BytePos,
-    ) -> Option<Ref<'_, BytePos, Vec<Comment>, FxBuildHasher>> {
-        self.leading.get(&pos)
+    fn add_trailing(&self, pos: BytePos, cmt: Comment) {
+        (**self).add_trailing(pos, cmt)
     }
 
-    pub fn move_leading(&self, from: BytePos, to: BytePos) {
-        let cmt = self.leading.remove(&from);
+    fn add_trailing_comments(&self, pos: BytePos, comments: Vec<Comment>) {
+        (**self).add_trailing_comments(pos, comments)
+    }
 
-        if let Some(cmt) = cmt {
-            self.leading.entry(to).or_default().extend(cmt.1);
-        }
+    fn has_trailing(&self, pos: BytePos) -> bool {
+        (**self).has_trailing(pos)
+    }
+
+    fn move_trailing(&self, from: BytePos, to: BytePos) {
+        (**self).move_trailing(from, to)
+    }
+
+    fn take_trailing(&self, pos: BytePos) -> Option<Vec<Comment>> {
+        (**self).take_trailing(pos)
+    }
+}
+
+impl<T> Comments for Rc<T>
+where
+    T: ?Sized + Comments,
+{
+    fn add_leading(&self, pos: BytePos, cmt: Comment) {
+        (**self).add_leading(pos, cmt)
     }
 
-    pub fn move_trailing(&self, from: BytePos, to: BytePos) {
-        let cmt = self.trailing.remove(&from);
+    fn add_leading_comments(&self, pos: BytePos, comments: Vec<Comment>) {
+        (**self).add_leading_comments(pos, comments)
+    }
+
+    fn has_leading(&self, pos: BytePos) -> bool {
+        (**self).has_leading(pos)
+    }
+
+    fn move_leading(&self, from: BytePos, to: BytePos) {
+        (**self).move_leading(from, to)
+    }
+
+    fn take_leading(&self, pos: BytePos) -> Option<Vec<Comment>> {
+        (**self).take_leading(pos)
+    }
+
+    fn add_trailing(&self, pos: BytePos, cmt: Comment) {
+        (**self).add_trailing(pos, cmt)
+    }
+
+    fn add_trailing_comments(&self, pos: BytePos, comments: Vec<Comment>) {
+        (**self).add_trailing_comments(pos, comments)
+    }
+
+    fn has_trailing(&self, pos: BytePos) -> bool {
+        (**self).has_trailing(pos)
+    }
+
+    fn move_trailing(&self, from: BytePos, to: BytePos) {
+        (**self).move_trailing(from, to)
+    }
+
+    fn take_trailing(&self, pos: BytePos) -> Option<Vec<Comment>> {
+        (**self).take_trailing(pos)
+    }
+}
+
+impl<T> Comments for Box<T>
+where
+    T: ?Sized + Comments,
+{
+    fn add_leading(&self, pos: BytePos, cmt: Comment) {
+        (**self).add_leading(pos, cmt)
+    }
+
+    fn add_leading_comments(&self, pos: BytePos, comments: Vec<Comment>) {
+        (**self).add_leading_comments(pos, comments)
+    }
+
+    fn has_leading(&self, pos: BytePos) -> bool {
+        (**self).has_leading(pos)
+    }
+
+    fn move_leading(&self, from: BytePos, to: BytePos) {
+        (**self).move_leading(from, to)
+    }
+
+    fn take_leading(&self, pos: BytePos) -> Option<Vec<Comment>> {
+        (**self).take_leading(pos)
+    }
```

Comment:
Would it be possible to have `get_leading(&self, pos: BytePos) -> Option<&[&Comment]>` and respective method for trailing ones?

Reply:
I expect it to require another complex generic traits.

But api like 

```rust
fn with_leading<F, Ret>(&self, op: F) -> Ret where F: FnOnce(&[Comment]) -> Ret;

```

is possible, regardless of threadedness

Reply:
Great, that would work for my use case 👍